agent/src/main/scala/za/co/absa/atum/agent/AtumAgent.scala

```diff
@@ -47,25 +47,41 @@ class AtumAgent private() {
    *  @return
    */
   def getOrCreateAtumContext(atumPartitions: AtumPartitions): AtumContext = {
-    contexts.getOrElse(atumPartitions, new AtumContext(atumPartitions, this))
+    getContextOrElse(
+      atumPartitions,
+      AtumContext.fromDTO(
+        dispatcher.getOrCreateAtumContext(AtumPartitions.toPartitioning(atumPartitions), None),
+        this
+      )
+    )
   }
 
   def getOrCreateAtumSubContext(subPartitions: AtumPartitions)(implicit parentAtumContext: AtumContext): AtumContext = {
     val newPartitions: AtumPartitions = parentAtumContext.atumPartitions ++ subPartitions
-    getContextOrElse(newPartitions, parentAtumContext.copy(atumPartitions = newPartitions, agent = this))
+    getContextOrElse(
+      newPartitions,
+      AtumContext.fromDTO(
+        dispatcher.getOrCreateAtumContext(
+          AtumPartitions.toPartitioning(newPartitions),
+          Some(AtumPartitions.toPartitioning(parentAtumContext.atumPartitions))
+        ),
+        this
+      )
+    )
   }
 
-  private def getContextOrElse(atumPartitions: AtumPartitions, creationMethod: =>AtumContext): AtumContext = {
-    synchronized{
-      contexts.getOrElse(atumPartitions, {
-        val result = creationMethod
-        contexts = contexts + (atumPartitions -> result)
-        result
-      })
+  private def getContextOrElse(atumPartitions: AtumPartitions, creationMethod: => AtumContext): AtumContext = {
```

Comment:
I think that the `creationMethod` is not a good name - it's actually a context already. If I understood your code and intentions correctly, then you are first calling the method and then its result is being passed here - so this is actually something like `existingContext` ?

Reply:
Refactored.